# Data Cleaning

## Load Data

######  Importation of the necessary libraries.

In [1]:
import os
import pandas as pd

##### Creation of a function which import the datasets of each month. Then it merges and saves them.

In [2]:
def merge_data(path, file_name):

    files = [file for file in os.listdir(path)]
    all_data = pd.DataFrame()
    for file in files:
        temp_file = pd.read_csv(path+'/'+file)
        all_data = pd.concat([all_data, temp_file])
    all_data.to_csv(file_name, index = False)

##### Utilisation of the function.

In [3]:
merge_data('./data','all_data.csv')

##### Checking of the new file.

In [4]:
df = pd.read_csv('all_data.csv')
df.head(3)

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   

       Order Date                   Purchase Address  
0  04/19/19 08:46       917 1st St, Dallas, TX 75001  
1             NaN                                NaN  
2  04/07/19 22:30  682 Chestnut St, Boston, MA 02215

## Exploration of the dataset

##### A quick exploration of the dataset before the cleaning.

In [5]:
def exploration(df):
    
    print(f'>SHAPE: {df.shape}')
    print('-' * 100)
    print(f'>COLUMN NAMES: {list(df.columns)}')
    print('-' * 100)
    print(f'>DESCRIBE: {df.describe()}')
    print('-' * 100)
    print(f'>ISNULL: {df.isnull().sum()}')
    print('-' * 100)
    print(f'>DTYPES: {df.dtypes}')

In [6]:
exploration(df)

>SHAPE: (186850, 6)
----------------------------------------------------------------------------------------------------
>COLUMN NAMES: ['Order ID', 'Product', 'Quantity Ordered', 'Price Each', 'Order Date', 'Purchase Address']
----------------------------------------------------------------------------------------------------
>DESCRIBE:         Order ID               Product Quantity Ordered Price Each  \
count     186305                186305           186305     186305   
unique    178438                    20               10         24   
top     Order ID  USB-C Charging Cable                1      11.95   
freq         355                 21903           168552      21903   

        Order Date  Purchase Address  
count       186305            186305  
unique      142396            140788  
top     Order Date  Purchase Address  
freq           355               355  
----------------------------------------------------------------------------------------------------
>ISNULL: Orde

##### Visualisation of the first and last three rows.

In [7]:
df.head(3)

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   

       Order Date                   Purchase Address  
0  04/19/19 08:46       917 1st St, Dallas, TX 75001  
1             NaN                                NaN  
2  04/07/19 22:30  682 Chestnut St, Boston, MA 02215

In [8]:
df.head(3)

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
1      NaN                         NaN              NaN        NaN   
2   176559  Bose SoundSport Headphones                1      99.99   

       Order Date                   Purchase Address  
0  04/19/19 08:46       917 1st St, Dallas, TX 75001  
1             NaN                                NaN  
2  04/07/19 22:30  682 Chestnut St, Boston, MA 02215

##### Visualisation of the products which are sold and the prices.

In [9]:
df['Product'].unique()

array(['USB-C Charging Cable', nan, 'Bose SoundSport Headphones',
       'Google Phone', 'Wired Headphones', 'Macbook Pro Laptop',
       'Lightning Charging Cable', '27in 4K Gaming Monitor',
       'AA Batteries (4-pack)', 'Apple Airpods Headphones',
       'AAA Batteries (4-pack)', 'iPhone', 'Flatscreen TV',
       '27in FHD Monitor', '20in Monitor', 'LG Dryer', 'ThinkPad Laptop',
       'Vareebadd Phone', 'LG Washing Machine', '34in Ultrawide Monitor',
       'Product'], dtype=object)

In [10]:
df['Price Each'].unique()

array(['11.95', nan, '99.99', '600', '11.99', '1700', '14.95', '389.99',
       '3.84', '150', '2.99', '700', '300', '149.99', '109.99', '600.0',
       '999.99', '400', '379.99', 'Price Each', '700.0', '1700.0',
       '150.0', '300.0', '400.0'], dtype=object)

## Data Cleaning

##### For each column there are 545 values that are NaN.

In [11]:
# How are distributed the NaN?
print(f'>SHAPE: {df[df.isnull().any(axis = 1)].shape}')
df[df.isnull().any(axis = 1)].head(3)

>SHAPE: (545, 6)


Order ID Product Quantity Ordered Price Each Order Date Purchase Address
1        NaN     NaN              NaN        NaN        NaN              NaN
356      NaN     NaN              NaN        NaN        NaN              NaN
735      NaN     NaN              NaN        NaN        NaN              NaN

##### We see that the NaN values are in 545 rows.

In [12]:
# What is the ratio of wrong rows?
print(f'>RATIO: {df[df.isnull().any(axis = 1)].shape[0]/df.shape[0]}')

>RATIO: 0.002916778164302917


##### We can delete these rows without losing  a lot of information.

In [13]:
df.dropna(inplace = True)
# We check the result.
df.isnull().sum()

Order ID            0
Product             0
Quantity Ordered    0
Price Each          0
Order Date          0
Purchase Address    0
dtype: int64

##### We see that the types of each column are objects, that is wrong. Thus we change this to the right type.

In [14]:
df.dtypes

Order ID            object
Product             object
Quantity Ordered    object
Price Each          object
Order Date          object
Purchase Address    object
dtype: object

In [15]:
df.head(3)

Order ID                     Product Quantity Ordered Price Each  \
0   176558        USB-C Charging Cable                2      11.95   
2   176559  Bose SoundSport Headphones                1      99.99   
3   176560                Google Phone                1        600   

       Order Date                      Purchase Address  
0  04/19/19 08:46          917 1st St, Dallas, TX 75001  
2  04/07/19 22:30     682 Chestnut St, Boston, MA 02215  
3  04/12/19 14:38  669 Spruce St, Los Angeles, CA 90001

In [16]:
df['Quantity Ordered'].unique()

array(['2', '1', '3', '5', 'Quantity Ordered', '4', '7', '6', '8', '9'],
      dtype=object)

In [17]:
print('>SHAPE: ',df[~df['Quantity Ordered'].str.isnumeric()].shape)

>SHAPE:  (355, 6)


##### We can see that 355 rows have not a numeric value in the column Quantity Ordered. We have to delete them.

In [18]:
df = df[df['Quantity Ordered'].str.isnumeric()]

In [19]:
# We don't need to do the same, because it has be done above.
df['Price Each'].unique()

array(['11.95', '99.99', '600', '11.99', '1700', '14.95', '389.99',
       '3.84', '150', '2.99', '700', '300', '149.99', '109.99', '600.0',
       '999.99', '400', '379.99', '700.0', '1700.0', '150.0', '300.0',
       '400.0'], dtype=object)

In [20]:
df['Quantity Ordered'] = pd.to_numeric(df['Quantity Ordered'])

In [21]:
df['Price Each'] = pd.to_numeric(df['Price Each'])

In [22]:
df['Order Date'] = pd.to_datetime(df['Order Date'])

In [23]:
df.dtypes

Order ID                    object
Product                     object
Quantity Ordered             int64
Price Each                 float64
Order Date          datetime64[ns]
Purchase Address            object
dtype: object

In [24]:
df.head()

Order ID                     Product  Quantity Ordered  Price Each  \
0   176558        USB-C Charging Cable                 2       11.95   
2   176559  Bose SoundSport Headphones                 1       99.99   
3   176560                Google Phone                 1      600.00   
4   176560            Wired Headphones                 1       11.99   
5   176561            Wired Headphones                 1       11.99   

           Order Date                      Purchase Address  
0 2019-04-19 08:46:00          917 1st St, Dallas, TX 75001  
2 2019-04-07 22:30:00     682 Chestnut St, Boston, MA 02215  
3 2019-04-12 14:38:00  669 Spruce St, Los Angeles, CA 90001  
4 2019-04-12 14:38:00  669 Spruce St, Los Angeles, CA 90001  
5 2019-04-30 09:27:00     333 8th St, Los Angeles, CA 90001

## Augmentation of the data

##### We split the purchase address in Address, City, State and the ZIP code.

In [25]:
def get_address(address):
    
    return address.split(',')[0]

def get_city(address):
    
    return address.split(',')[1]

def get_state(address):
    
    temp = address.split(',')[2]
    return temp.split(' ')[1]

def get_zip(address):
    
    temp = address.split(',')[2]
    return temp.split(' ')[2]

In [26]:
df['Address'] = df['Purchase Address'].apply(lambda x: get_address(x))
df['City'] = df['Purchase Address'].apply(lambda x: get_city(x))
df['State'] = df['Purchase Address'].apply(lambda x: get_state(x))
df['ZIP'] = df['Purchase Address'].apply(lambda x: get_zip(x))

In [27]:
df.drop('Purchase Address', axis = 1, inplace = True)

In [28]:
df.head(3)

Order ID                     Product  Quantity Ordered  Price Each  \
0   176558        USB-C Charging Cable                 2       11.95   
2   176559  Bose SoundSport Headphones                 1       99.99   
3   176560                Google Phone                 1      600.00   

           Order Date          Address          City State    ZIP  
0 2019-04-19 08:46:00       917 1st St        Dallas    TX  75001  
2 2019-04-07 22:30:00  682 Chestnut St        Boston    MA  02215  
3 2019-04-12 14:38:00    669 Spruce St   Los Angeles    CA  90001